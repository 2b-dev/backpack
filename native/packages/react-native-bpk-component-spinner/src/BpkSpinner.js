/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2017 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import PropTypes from 'prop-types';
import { ActivityIndicator } from 'react-native';
import { withTheme } from 'react-native-bpk-theming';
import { colorBlue500, colorWhite, colorGray700 } from 'bpk-tokens/tokens/base.react.native';

const SPINNER_TYPES = ['primary', 'light', 'dark'];

// If theming is ever expanded to support other types, this should be changed
// to something akin to BpkButton's theming functions.
const THEMING_ATTRIBUTE = 'spinnerPrimaryColor';

const themePropType = (props, propName, componentName) => {
  const { type, theme } = props;
  if (!theme) {
    return false;
  }
  if (type === 'primary' && !theme[THEMING_ATTRIBUTE]) {
    return new Error(`Invalid prop \`${propName}\` supplied to \`${componentName}\`. For spinners of type \`${type}\`, the \`theme\` prop must include \`${THEMING_ATTRIBUTE}\``); // eslint-disable-line max-len
  }
  return false;
};

const getSpinnerColor = (theme, type) => {
  const colorMappings = {
    primary: colorBlue500,
    light: colorWhite,
    dark: colorGray700,
  };
  if (theme && type === 'primary') {
    return theme[THEMING_ATTRIBUTE];
  }
  return colorMappings[type];
};

const BpkSpinner = (props) => {
  const { small, type, ...rest } = props;
  let { theme } = props;

  // Validate type.
  if (!SPINNER_TYPES.includes(type)) {
    throw new Error(`"${type}" is not a valid spinner type. Valid types are ${SPINNER_TYPES.join(', ')}`);
  }

  // Validate that spinner is themeable and correct theming attribute has been
  // supplied. If not, disable theming.
  if (theme && (type !== 'primary' || !theme[THEMING_ATTRIBUTE])) {
    theme = null;
  }

  return (
    <ActivityIndicator
      color={getSpinnerColor(theme, type)}
      size={small ? 'small' : 'large'}
      {...rest}
    />
  );
};

const propTypes = {
  small: PropTypes.bool,
  theme: themePropType,
  type: PropTypes.oneOf(SPINNER_TYPES),
};

BpkSpinner.propTypes = propTypes;

BpkSpinner.defaultProps = {
  small: false,
  theme: null,
  type: 'primary',
};

export default withTheme(BpkSpinner);
export {
  propTypes,
  SPINNER_TYPES,
};
