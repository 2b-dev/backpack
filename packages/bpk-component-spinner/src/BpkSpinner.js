import React, { PropTypes } from 'react'
import TOKENS from 'bpk-tokens/tokens/base.common'

import './bpk-spinner.scss'

const BpkSpinner = ({ large, extraLarge, color }) => {
  const classNames = [ 'bpk-spinner' ]
  const __html = require('raw!bpk-svgs/src/spinners/spinner.svg')
    .replace(/^<svg/, `<svg fill="${color}"`)

  large ? classNames.push('bpk-spinner--large') : null
  extraLarge ? classNames.push('bpk-spinner--extra-large') : null

  return <span className={classNames.join(' ')} dangerouslySetInnerHTML={{ __html }} />
}

BpkSpinner.propTypes = {
  large: PropTypes.bool,
  extraLarge: PropTypes.bool,
  color: PropTypes.string
}

BpkSpinner.defaultProps = {
  large: false,
  extraLarge: false,
  color: TOKENS.colorGray500
}

export default BpkSpinner
