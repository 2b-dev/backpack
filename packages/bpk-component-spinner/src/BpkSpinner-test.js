import React from 'react'
import renderer from 'react/lib/ReactTestRenderer'
import BpkSpinner from './BpkSpinner'

describe('BpkSpinner', () => {
  it('should render correctly', () => {
    const tree = renderer.create(<BpkSpinner />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should render correctly with a "alignToButton" attribute', () => {
    const tree = renderer.create(<BpkSpinner alignToButton />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should render correctly with a "large" attribute', () => {
    const tree = renderer.create(<BpkSpinner large />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should render correctly with a "large" and a "alignToButton" attribute', () => {
    const tree = renderer.create(<BpkSpinner large alignToButton />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should render correctly with a "extraLarge" attribute', () => {
    const tree = renderer.create(<BpkSpinner extraLarge />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
