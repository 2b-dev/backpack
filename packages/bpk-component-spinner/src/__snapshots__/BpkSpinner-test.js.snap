exports[`BpkSpinner should render correctly 1`] = `
<span
  className="bpk-spinner"
  dangerouslySetInnerHTML={
    Object {
      "__html": "FILE_MOCK_CONTENTS"
    }
  } />
`;

exports[`BpkSpinner should render correctly with a "extraLarge" attribute 1`] = `
<span
  className="bpk-spinner bpk-spinner--extra-large"
  dangerouslySetInnerHTML={
    Object {
      "__html": "FILE_MOCK_CONTENTS"
    }
  } />
`;

exports[`BpkSpinner should render correctly with a "large" attribute 1`] = `
<span
  className="bpk-spinner bpk-spinner--large"
  dangerouslySetInnerHTML={
    Object {
      "__html": "FILE_MOCK_CONTENTS"
    }
  } />
`;
